apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,205 @@
+        {
+            // If we are not retrying failed calls either propagate or swallow the exception.
+            if (_options.RetryType == RetryType.None)
+            {
+                switch (_options.ExceptionHandling)
+                {
+                    case ExceptionHandling.Propagate:
+                        return false;
+                    case ExceptionHandling.Ignore:
+                        Debug.WriteLine($"Ignored exception: {e}");
+                        return true;
+                    default:
+                        Debug.Fail($"Invalid ExceptionHandling option: {_options.ExceptionHandling}");
+                        return true;
+                }
+            }
+            
+            // Get the size of items to be buffered and retried.
+            int size = items.Sum(_sizer);
+
+            // Lock the semaphore and add the new items to the buffer.
+            lock (_mutex)
+            {
+                // If the new items would be greater than the size of the buffer drop them or throw.
+                if (size + _sizeInBytes > _options.BufferSizeBytes) 
+                {
+                    switch (_options.BufferOverflow)
+                    {
+                        case BufferOverflow.IgnoreNewEntries:
+                            Debug.WriteLine($"Ignored {items.Count()} entries.");
+                            return true;
+                        case BufferOverflow.ThrowException:
+                            throw new InvalidOperationException("Buffer overflow", e);
+                        default:
+                            Debug.Fail($"Invalid BufferOverflow option: {_options.BufferOverflow}");
+                            return true;
+                    }
+                }
+            
+                _sizeInBytes += size;
+                _buffer.AddRange(items);
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Try to flush the buffer if it has any entries.
+        /// </summary>
+        private void FlushBuffer()
+        {
+            // Move the current buffer into a new buffer to send.
+            List<T> toSend;
+            int toSendSize;
+            lock (_mutex)
+            {
+                if (_buffer.Count() == 0)
+                {
+                    return;
+                }
+                toSend = _buffer;
+                toSendSize = _sizeInBytes;
+                _buffer = new List<T>();
+                _retries++;
+            }
+
+            try
+            {
+                _consumer.Receive(toSend);
+                lock (_mutex)
+                {
+                    _retries = 0;
+                    _sizeInBytes -= toSendSize;
+                }
+            }
+            catch (RpcException) 
+            {
+                if (_retries >= _options.RetryAttempts)
+                {
+                    Interlocked.Exchange(ref _retries, 0);
```

Comment:
Why do we reset the number of retries here? Doesn't that mean we'll keep retrying? It's not really clear to me what the behavior should be at this point if we've said we want to swallow exceptions. Am I right in saying that at the moment, it's "drop the contents of the buffer, but retry future buffers"?